### PR TITLE
Subscriptions

### PR DESCRIPTION
**Pull request by kngako**
https://github.com/stakwork/sphinx-kotlin/issues/296

### CODE REVIEW

**Comment by tomastiminskas** [comment]
@05nelsonm this PR as the #302 includes a db migration with name 2.sqm so once #302 is approved and merged, we should merge develop into this branch and rename the migration to 3.sqm

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionFragment.kt [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                    endNumber
+                )
+            }
+            (requireActivity() as InsetterActivity).addNavigationBarPadding(layoutConstraintSubscription)
+        }
+
+        viewModel.initSubscription()
+    }
+
+    override suspend fun onViewStateFlowCollect(viewState: SubscriptionViewState) {
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionFragment.kt:166]: The primary purpose of a `ViewState` is to describe the state of the view, not only hold the data and then based off of the data, set the view. As it looks now, one cannot look at the xml layout, and the view state, and decipher what each view's state is. All of the logic is in the renderer when it should belong in the ViewModel (or when the view's state is instantiated for immutability/encapsulation purposes).

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                                        subscription.end_number,
+                                        subscription.count,
+                                        subscription.end_date,
+                                        subscription.ended,
+                                        paused = true,
+                                        subscription.created_at,
+                                        subscription.updated_at,
+                                        subscription.chat_id,
+                                        subscription.contact_id
+                                    )
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt:197]: unnecessary. Remove.

success case returns a paused, updated subscription object and can be set w/o need to re-instantiate a `Subscription` object.

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                                        subscription.end_number,
+                                        subscription.count,
+                                        subscription.end_date,
+                                        subscription.ended,
+                                        paused = false,
+                                        subscription.created_at,
+                                        subscription.updated_at,
+                                        subscription.chat_id,
+                                        subscription.contact_id
+                                    )
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt:237]: unnecessary. Remove and use returned sub.

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewState.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+internal sealed class SubscriptionViewState: ViewState<SubscriptionViewState>() {
+    object Idle: SubscriptionViewState()
+
+    object SavingSubscription: SubscriptionViewState()
+
+    object SavingSubscriptionFailed: SubscriptionViewState()
+
+    class SubscriptionLoaded(
+        val subscription: Subscription
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewState.kt:14]: subscription data does not describe the view's state. move all logic to view model.

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/widgets/SphinxRadioGroup.kt [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.core.content.ContextCompat
+import chat.sphinx.subscription.ui.widgets.SphinxRadioGroup
+
+class SphinxRadioGroup @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0,
+    defStyleRes: Int = 0
+) : ConstraintLayout(context, attrs, defStyleAttr, defStyleRes) {
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/widgets/SphinxRadioGroup.kt:23]: Remove Use of `SphinxRadioGroup` custom class.

Custom views such as this:
 - Lock things to the Android platform
 - Does not promote extensibility
 - Needs testing
 - Couples the View and State

This functionality could easily be achieved using the rest of the application's already implemented MVI pattern (`ViewState`) to describe the state of the view, and update it upon reaction to events (such as when a user clicks 1 check box, the rest of them on the view state would change to `checked = false`). Utilizing `ViewState` allows for use of primitive views (`ConstraintLayout`, `CheckBox`, `TextView`) that are already very well tested such that we can de-couple the state from the View by simply rendering things based off of what is described.

Making views and fragments as "dumb" as possible is the primary goal.

**Review by 05nelsonm — CHANGES_REQUESTED** [review]

File: sphinx/application/data/features/feature-repository/src/main/java/chat/sphinx/feature_repository/SphinxRepository.kt [hunk 6]
```diff
@@ -3550,4 +3561,237 @@ abstract class SphinxRepository(
...
+                .map { it?.let { subscriptionDboPresenterMapper.mapFrom(it) } }
+                .distinctUntilChanged()
+        )
+    }
+
+    override suspend fun getActiveSubscriptionByContactId(contactId: ContactId): Flow<Subscription?> = flow {
+        emitAll(
+            coreDB.getSphinxDatabaseQueries().activeSubscriptionGetByContactId(contactId)
+                .asFlow()
+                .mapToOneOrNull(io)
```
[the remark this answers is not in the document]
Comment (tomastiminskas) [reply sphinx/application/data/features/feature-repository/src/main/java/chat/sphinx/feature_repository/SphinxRepository.kt:3578]: Was changed and renamed to return last active subscription by contact id

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt [hunk 7]
```diff
@@ -39,23 +41,40 @@ internal class SubscriptionViewModel @Inject constructor(
...
+    companion object {
+        const val DAILY_INTERVAL: String = "daily"
+        const val WEEKLY_INTERVAL: String = "weekly"
+        const val MONTHLY_INTERVAL: String = "MONTHLY"
+    }
+
     fun initSubscription() {
         viewModelScope.launch(mainImmediate) {
-            subscriptionRepository.getActiveSubscriptionByContactId(ContactId(args.argContactId)).firstOrNull().let { subscription ->
+            subscriptionRepository.getActiveSubscriptionByContactId(ContactId(args.argContactId)).distinctUntilChanged().collect { subscription ->
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt:52]: this change now results in a never ending collection of the subscription that is initialized every time the fragment view goes through the `onViewCreated()` resulting in multiple coroutines collecting. Needs to be moved to a custom view state container which utilizes a flow builder and `.stateIn()` extension function with WhileSubscribed in order to cancel sub when screen's lifecycle moves to the `onStop` event.

**Review by 05nelsonm — CHANGES_REQUESTED** [review]

File: sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt [hunk 8]
```diff
@@ -47,36 +47,56 @@ internal class SubscriptionViewModel @Inject constructor(
...
-                    } else {
-                        WEEKLY_INTERVAL
-                    }
+    private val subscriptionSharedFlow: SharedFlow<Subscription?> = flow {
+        emitAll(subscriptionRepository.getActiveSubscriptionByContactId(ContactId(args.argContactId)))
+    }.shareIn(
+        viewModelScope,
+        SharingStarted.WhileSubscribed(2_000),
+        replay = 1,
+    )
```
Comment (05nelsonm) [inline sphinx/screens-detail/subscription/subscription/src/main/java/chat/sphinx/subscription/ui/SubscriptionViewModel.kt:56]: remove.

not necessary.

**Review by 05nelsonm — CHANGES_REQUESTED** [review]

**Review by 05nelsonm — APPROVED** [review]

**Pull request merged**